### PR TITLE
Stlx 14208 refactor near builder

### PR DESCRIPTION
**Pull request by facundoCarreSimtlix**
refactor transaction builder

### CODE REVIEW

File: modules/account-lib/src/coin/near/transactionBuilder.ts [hunk 1]
```diff
@@ -135,29 +135,65 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
...
 
-  public sender(value: string) {
-    this._sender = value;
+  /**
+   * Sets the public key of the sender of this transaction.
+   *
+   * @param {string} senderAddress the public key of the account that is sending this transaction
+   * @returns {TransactionBuilder} This transaction builder
+   */
+  public sender(publicKey: string): this {
```
Comment (Krin2) [inline modules/account-lib/src/coin/near/transactionBuilder.ts:144]: I think publicKey should be change to senderAddress

File: modules/account-lib/src/coin/near/transactionBuilder.ts [hunk 2]
```diff
@@ -135,29 +135,65 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
...
+    if (!utils.isValidAddress(publicKey)) {
+      throw new BuildTransactionError(`Invalid sender public key ${publicKey}`);
+    }
+    this._sender = publicKey;
+    return this;
+  }
+
+  /**
+   * Sets the public key of the sender of this transaction.
```
Comment (Krin2) [inline modules/account-lib/src/coin/near/transactionBuilder.ts:153]: change jsdoc

File: modules/account-lib/src/coin/near/transactionBuilder.ts [hunk 3]
```diff
@@ -135,29 +135,65 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
...
+   *
+   * @param {string} senderAddress the public key of the account that is sending this transaction
```
Comment (Krin2) [inline modules/account-lib/src/coin/near/transactionBuilder.ts:155]: senderAddress should be accountId

File: modules/account-lib/src/coin/near/transactionBuilder.ts [hunk 4]
```diff
@@ -135,29 +135,65 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
...
+   */
+  public senderAccountId(accountId: string): this {
+    utils.isValidAddress(accountId);
+    this._senderAccountId = accountId;
+    return this;
   }
 
-  public receiverId(value: string) {
-    this._receiverId = value;
+  /** TODO: add jsdocs */
```
Comment (Krin2) [inline modules/account-lib/src/coin/near/transactionBuilder.ts:164]: Add jsdoc

File: modules/account-lib/src/coin/near/transactionBuilder.ts [hunk 5]
```diff
@@ -135,29 +135,65 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
...
+    this._receiverId = accountId;
+    return this;
   }
 
-  public nounce(value: number) {
-    this._nonce = value;
+  /**
+   * Set the nonce
+   *
+   * @param {nonce} number - number that can be only used once
```
Comment (Krin2) [inline modules/account-lib/src/coin/near/transactionBuilder.ts:174]: `@param {number} nonce`

File: modules/account-lib/src/coin/near/transactionBuilder.ts [hunk 6]
```diff
@@ -135,29 +135,65 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
...
+    if (nonce < 0) {
+      throw new BuildTransactionError(`Invalid nonce: ${nonce}`);
+    }
+    this._nonce = nonce;
+    return this;
   }
 
-  public recentBlockHash(value: string) {
-    this._recentBlockHash = value;
+  //* * TODO: add jsdocs */
```
Comment (Krin2) [inline modules/account-lib/src/coin/near/transactionBuilder.ts:185]: add jsdoc

**Review by ipelaez-simtlix — COMMENTED** [review]
Please, lets check this, we have changed this value ( swap ) and now some values must be swaped also at the moment of build things, for example receive fisrt the signerId and second the public key but now is first public key then signerId, lets fix this and make sure that its not happening on other part of the code.

    const tx = nearAPI.transactions.createTransaction(
      this._sender,
      nearAPI.utils.PublicKey.fromString(nearAPI.utils.serialize.base_encode(hex.decode(this._senderAccountId))),
      
      PD: we need to test it with coin-sandbox, which is part of the story

File: modules/account-lib/src/coin/near/transactionBuilder.ts [hunk 7]
```diff
@@ -11,6 +11,10 @@ import assert from 'assert';
 import { KeyPair } from './keyPair';
 import * as hex from '@stablelib/hex';
 
+export interface SendParams {
```
Comment (rliberal-simtlix) [inline modules/account-lib/src/coin/near/transactionBuilder.ts:14]: Rename interface as SenderParams

**Review by rliberal-simtlix — DISMISSED** [review]

File: modules/account-lib/src/coin/near/transactionBuilder.ts [hunk 8]
```diff
@@ -11,6 +11,10 @@ import assert from 'assert';
 import { KeyPair } from './keyPair';
 import * as hex from '@stablelib/hex';
 
+export interface SenderParams {
+  address: string;
+  pubKey: string;
```
Comment (ewangbitgo) [inline modules/account-lib/src/coin/near/transactionBuilder.ts:16]: creating a interface for just combine to parameter seems over kill

File: modules/account-lib/src/coin/near/transactionBuilder.ts [hunk 9]
```diff
@@ -140,29 +144,75 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
...
 
-  public sender(value: string) {
-    this._sender = value;
+  /**
+   * Sets the public key and the address of the sender of this transaction.
+   *
+   * @param {SenderParams} publicKey the public key and account that is sending this transaction
+   * @returns {TransactionBuilder} This transaction builder
+   */
+  public sender({ address, pubKey }: SenderParams): this {
```
Comment (ewangbitgo) [inline modules/account-lib/src/coin/near/transactionBuilder.ts:153]: public sender(address: string, pubKey: string) seems simpler

**Review by ewangbitgo — CHANGES_REQUESTED** [review]

File: modules/account-lib/src/coin/near/transactionBuilder.ts (hunk 9 above)
Comment (facundoCarreSimtlix) [reply modules/account-lib/src/coin/near/transactionBuilder.ts:153]: from what I saw it,  always try to use  an interface for the parameters, but if you like i change it

Comment (rliberal-simtlix) [reply modules/account-lib/src/coin/near/transactionBuilder.ts:153]: lets rollback and use two parameters.

**Review by rliberal-simtlix — APPROVED** [review]

File: modules/account-lib/src/coin/near/transferBuilder.ts [hunk 10]
```diff
@@ -7,6 +7,7 @@ import * as NearAPI from 'near-api-js';
 import assert from 'assert';
 import { BuildTransactionError } from '../baseCoin/errors';
 import BN from 'bn.js';
+import BigNumber from 'bignumber.js';
```
Comment (Krin2) [inline modules/account-lib/src/coin/near/transferBuilder.ts:10]: I think we should use just one of this: BN or bignumber both work with Big Numbers

File: modules/account-lib/src/coin/near/transferBuilder.ts [hunk 11]
```diff
@@ -41,7 +42,15 @@ export class TransferBuilder extends TransactionBuilder {
     return tx;
   }
 
-  public amount(value: string) {
+  /**
+   * Sets the amount of this transaction.
+   *
+   * @param {string} value the the list of actions
```
Comment (Krin2) [inline modules/account-lib/src/coin/near/transferBuilder.ts:48]: correct description

File: modules/account-lib/src/coin/near/transferBuilder.ts (hunk 10 above)
Comment (rliberal-simtlix) [reply modules/account-lib/src/coin/near/transferBuilder.ts:10]: Please use bignumber.js

Comment (rliberal-simtlix) [reply modules/account-lib/src/coin/near/transferBuilder.ts:10]: near sdk https://www.npmjs.com/package/near-api-js uses bn.js not bignumber.js all interfaces uses that library

**Review by ewangbitgo — APPROVED** [review]

**Pull request merged**